File: 1710.rb
# @param {Integer[][]} box_types
# @param {Integer} truck_size
# @return {Integer}
def maximum_units(box_types, truck_size)
  total = 0
  box_types.sort_by {|i| -i[1]}.each do |boxes, units|
    total += units * (boxes <= truck_size ? boxes : truck_size)
    break if (truck_size -= boxes) <= 0
  end
  total
end

File: 1710.js
"use strict";

/**
 * @param {number[][]} boxTypes
 * @param {number} truckSize
 * @return {number}
 */
function maximumUnits(boxTypes, truckSize) {
    boxTypes.sort((a, b) => b[1] - a[1]);
    let total = 0;
    for (let [boxes, units] of boxTypes) {
        total += units * (boxes <= truckSize ? boxes : truckSize);
        truckSize -= boxes;
        if (truckSize <= 0)
            break;
    }
    return total;
}

console.log(maximumUnits([[1, 3], [2, 2], [3, 1]], 4) == 8);
console.log(maximumUnits([[5, 10], [2, 5], [4, 7], [3, 9]], 10) == 91);
